pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/MetadataStoreFactoryImpl.java

```diff
@@ -45,19 +62,50 @@ public static MetadataStoreExtended createExtended(String metadataURL, MetadataS
     private static MetadataStore newInstance(String metadataURL, MetadataStoreConfig metadataStoreConfig,
                                              boolean enableSessionWatcher)
             throws MetadataStoreException {
+        MetadataStoreProvider provider = findProvider(metadataURL);
+        if (provider != null) {
+            return provider.create(metadataURL, metadataStoreConfig, enableSessionWatcher);
+        }
+        return new ZKMetadataStore(metadataURL, metadataStoreConfig, enableSessionWatcher);
+    }
+
+    static void loadProviders() {
+        String factoryClasses = System.getProperty(METADATASTORE_PROVIDERS_PROPERTY);
```

Comment:
Why are you using a System Property and not the standard pulsar configuration, which can be loaded from a file and documented properly?

Reply:
This config needs to be loaded from many places, CLI tools to make it less practical to add the config option everywhere. We also need it in BookKeeper process and tools, where we have cannot add Pulsar's specific configs.

Reply:
@merlimat Which CLI tools need direct access to the Metadata store?
I only know ./pulsar-admin, which REST directly to the broker, so it doesn’t need meta.
Can you help me understand what the many places that don’t have access to pulsar configuration are?
Regarding BK: Isn’t BK a separate codebase, so it won’t share this code anyway?

Reply:
@asafm 

1. Based on the call hierarchy of `newInstance`, these are some of those tools/services:
```
CompactorTool.main(String[])  (org.apache.pulsar.compaction)
ManagedLedgerWriter.main(String[])  (org.apache.pulsar.testclient)
ProxyServiceStarter.main(String[])  (org.apache.pulsar.proxy.server)
PulsarClusterMetadataSetup.main(String[])  (org.apache.pulsar)
PulsarClusterMetadataTeardown.main(String[])  (org.apache.pulsar)
PulsarInitialNamespaceSetup.doMain(String[])  (org.apache.pulsar)
PulsarStandaloneStarter.main(String[])  (org.apache.pulsar)
PulsarTransactionCoordinatorMetadataSetup.main(String[])  (org.apache.pulsar)
WebSocketServiceStarter.main(String[])  (org.apache.pulsar.websocket.service)
```
3. Pulsar implements some BK interfaces to be able to plug its implementations as providers in the BK client and bookie. see https://github.com/apache/pulsar/blob/master/bin/pulsar#L353-L359